### PR TITLE
85 update analyzers to use the new trace call stack graph implementation

### PR DESCRIPTION
**Pull request by fengxizhou**
## What does this PR do?
Update the open source HTA analyzers to use the new trace call stack implementation. This PR makes the following updates:

1. Refactor `CuptiCounterAnalysis` to use the new Trace CallGraph implementation.
2. Update `CriticalPathAnalysis` to use the new implementation.
3. Update `get_frequent_cuda_kernel_sequences` to use the new implementation.
4. Add back `CallStackGraph.dfs_traverse()` method.
5. Correct the assertion on the value of `botton_level_op` in `test_get_cupti_counter_data_with_operators()`.
6. Update `infer_device_type()` to use `pid` and `id` columns in addition to the `stream` column.
7. Move the unit tests for `CuptiCounterAnalysis` and `CriticalPathAnalysis` to separate test files.

## Before submitting

- [X] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [X] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [] Did you update the [changelog](https://github.com/facebookresearch/HolisticTraceAnalysis/blob/main/CHANGELOG.md)?
  - [X] N/A

### CODE REVIEW

File: hta/analyzers/cupti_counter_analysis.py [hunk 1]
```diff
@@ -17,52 +17,36 @@
...
-        ].reset_index(drop=True)
-
-        # We map 1:1 each kernel launch with gpu kernel
-        # so the number of each should be equal
-
+        # Get cuda kernel launches
+        kernel_launches = (
+            df.loc[
+                df.s_cat.str.match("cuda_runtime")
+                & df.s_name.str.match("cudaLaunchKernel")
```
Comment (briancoutinho) [inline hta/analyzers/cupti_counter_analysis.py:47]:
```
             | (
                df.s_cat.str.match("cuda_drove")
                & df.s_name.str.match("cuLaunchKernel")
                )
```

File: hta/analyzers/cupti_counter_analysis.py [hunk 2]
```diff
@@ -72,60 +56,41 @@ def _get_counter_data_with_operators_for_rank(
...
+        gpu_kernels["index_runtime"] = kernel_launches["index"]
+        # Add stack columns
+        op_stacks = {}
+        top_level_ops = {}
+        bottom_level_ops = {}
+        for idx, kernel_idx, launch_idx in gpu_kernels[
+            ["index", "index_runtime"]
+        ].itertuples(index=True):
+            stack = cg.get_stack_of_node(launch_idx).sort_values("ts")
+            ops = stack.loc[stack.s_cat.eq("cpu_op"), "index"].to_list()
```
Comment (briancoutinho) [inline hta/analyzers/cupti_counter_analysis.py:68]: @fengxizhou  I thought stacks contain runtime events too, that is why on the left I skipped one level, that might be wrong assumption, please let me kno

File: hta/common/trace_call_stack.py [hunk 3]
```diff
@@ -871,3 +874,22 @@ def _dfs(idx: int) -> KernelInfo:
         logger.debug(
             f"Add kernel info: prepare {t1 - t0:.2}s; traverse in {t2 - t1:.2f}s; update df {t3 - t2:.2f}s"
         )
+
+    def dfs_traverse(self, enter_func: DFSCallback, exit_func: DFSCallback) -> None:
```
Comment (briancoutinho) [inline hta/common/trace_call_stack.py:878]: thanks for adding, lgtm :)

File: tests/test_critical_path_analysis.py [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+from collections import Counter
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Tuple
+
+from hta.analyzers.critical_path_analysis import CPEdge, CPEdgeType
+from hta.trace_analysis import TraceAnalysis
+
+
+class CriticalPathAnalysisTestCase(unittest.TestCase):
```
Comment (briancoutinho) [inline tests/test_critical_path_analysis.py:14]: thanks for seperating this out, assuming these are identical to previous test in test_trace_analysis so shoudl be good :)

File: tests/test_cupti_profiler.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        test_row = counters_df.sort_values(axis=0, by="ts").iloc[5].to_dict()
+
+        self.assertEqual(test_row["cat"], "cuda_profiler_range")
+        self.assertTrue("fft2d_r2c_32x32" in test_row["name"])
+
+        self.assertEqual(
+            test_row["smsp__sass_thread_inst_executed_op_fadd_pred_on.sum"], 86114304
+        )
+        self.assertEqual(test_row["top_level_op"], "aten::conv2d")
+        # self.assertEqual(test_row["bottom_level_op"], "aten::_convolution")
```
Comment (briancoutinho) [inline tests/test_cupti_profiler.py:45]: please see my comment in cupti_counter_analysis.py, if that looks good, this is great, thanks :)

**Review by briancoutinho — APPROVED** [review]
Thanks @fengxizhou  for updating this and adding improvements to the code.
Left a few minor comments and questino/clarifications.

File: hta/analyzers/cupti_counter_analysis.py (hunk 2 above)
Comment (fengxizhou) [reply hta/analyzers/cupti_counter_analysis.py:68]: Yes. It includes the runtime. However, here we only extract the `cpu_op` events.

File: tests/test_critical_path_analysis.py (hunk 4 above)
Comment (fengxizhou) [reply tests/test_critical_path_analysis.py:14]: They are the same. With separate test files, it easy to debug and test.

File: hta/analyzers/cupti_counter_analysis.py (hunk 1 above)
Comment (fengxizhou) [reply hta/analyzers/cupti_counter_analysis.py:47]: We can change line 46 to
 ```df.s_cat.str.match("(cuda_runtime)|(cuda_drove)")```

[the remark this answers is not in the document]
Comment (fengxizhou) [reply hta/analyzers/cupti_counter_analysis.py:47]: We can change line 47 to ```& df.s_name.str.match("cudaLaunch")```

**Pull request closed without merging**